src/target/riscv/riscv-013.c

```diff
@@ -2619,7 +2619,7 @@ static int read_memory_progbuf_inner(struct target *target, target_addr_t addres
 		/* Now read whatever we got out of the batch. */
 		dmi_status_t status = DMI_STATUS_SUCCESS;
 		for (size_t i = 0; i < reads; i++) {
-			riscv_addr_t receive_addr = read_addr + (i-2) * size;
+			riscv_addr_t receive_addr = read_addr + ((int64_t)i-2) * size;
```

Comment:
I think a better fix is to change the type of `i` to riscv_addr_t. Does that work?